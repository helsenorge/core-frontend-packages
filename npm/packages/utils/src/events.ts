export default {
  on(
    el: Document | HTMLElement | Element,
    type: string,
    callback: (event?: MouseEvent | React.MouseEvent<{}>) => void
  ): void {
    if (el.addEventListener) {
      el.addEventListener(type, callback as EventListener);
    }
  },

  off(
    el: Document | HTMLElement | Element,
    type: string,
    callback: (event?: MouseEvent | React.MouseEvent<{}>) => void
  ): void {
    if (el.removeEventListener) {
      el.removeEventListener(type, callback as EventListener);
    }
  }
};
