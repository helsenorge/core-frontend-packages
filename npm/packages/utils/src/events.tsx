/* This is a custom interface to be able to access and typecheck path on Events */
export interface WebComponentEvent<T extends EventTarget> extends Event {
  target: T;
  tagName: string;
  path: {
    find: (fn: (e: WebComponentEvent<T>) => boolean) => { id: string };
  };
}

/* This is a util to get the event target, wether it comes from the lightDOM or shadowDOM - this requires the Shadow DOM to be of type 'open' */
export const getEventTarget = (e: Event): EventTarget => e.composedPath()[0];
