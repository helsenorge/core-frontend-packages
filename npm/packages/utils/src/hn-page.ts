/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
declare const HN: any;

window.HN = window.HN || {};
window.HN.MinHelse = window.HN.MinHelse || {};
window.HN.Page = window.HN.Page || {};

export type callback = () => void;

/**
 * Returnerer __Path__ fra HN.Page objektet
 */
export const getPath = (): string => {
  return HN.Page.__Path__;
};
/**
 * Returnerer __Version__ fra HN.Page objektet
 */
export const getVersion = (): string => {
  return HN.Page.__Version__;
};
/**
 * Returnerer __Assets__ fra HN.Page objektet
 */
export const getAssets = (): string => {
  return HN.Page.__Assets__;
};

/**
 * Returnerer __Assets__ fra HN.Page objektet - eller tom string hvis den ikke finnes
 */
export const getAssetsUrl = (): string => {
  return HN.Page.__Assets__ !== undefined && HN.Page.__Assets__ !== null ? HN.Page.__Assets__ : '';
};

/**
 * Oppdaterer __Path__ key på HN.Page objektet
 * @param newPath verdien som settes i __Path__
 */
export const setPath = (newPath: string): void => {
  HN.Page.__Path__ = newPath;
};

/**
 * Sjekker eksisterende Queue på HN.MinHelse objektet og pusher cb methoden i Queuen
 * @param start callback som pushes i Queue
 */
export const registerBlock = (start: callback): void => {
  if (!window.HN.MinHelse.hasOwnProperty('Queue')) {
    window.HN.MinHelse.Queue = [];
    start();
  } else {
    const queue: Array<callback> = window.HN.MinHelse.Queue as Array<callback>;
    queue.push(start);
  }
};

/**
 * Sjekker eksisterende Queue på HN.MinHelse objektet og pusher cb methoden i Queuen
 * Forskjell med registerBlock er at den  oppretter en tom array i Queue hvis den ikke eksisterer fra før av
 * @param start callback som pushes i Queue
 */
export const register = (start: callback): void => {
  if (!window.HN.MinHelse.hasOwnProperty('Queue')) {
    start();
  } else {
    const queue: Array<callback> = window.HN.MinHelse.Queue as Array<callback>;
    queue.push(start);
  }
};
