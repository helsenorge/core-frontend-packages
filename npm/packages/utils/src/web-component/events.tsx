import { useEffect } from 'react';

import * as History from 'history';

import { ShowSignOutBoxData, SessionTimeoutAction } from '../types/entities';

/* Events from header-footer and cms-blocks webcomp */
import { setVisPersonvelger } from '../hn-user';
import { HeaderFooterEvents, CmsBlocksEvents, CommonEvents } from './constants';

export interface KeyboardEventWithPath extends KeyboardEvent {
  path?: string;
}
export interface MouseEventWithPath extends MouseEvent {
  path?: string;
}

/* This is a custom interface to be able to access and typecheck path on Events */
export interface WebComponentEvent<T extends EventTarget> extends Event {
  target: T;
  tagName: string;
  path: {
    find: (fn: (e: WebComponentEvent<T>) => boolean) => { id: string };
  };
}

/* This is a util to get the event target, wether it comes from the lightDOM or shadowDOM - this requires the Shadow DOM to be of type 'open' */
export const getEventTarget = (e: Event | MouseEvent | KeyboardEvent): EventTarget => {
  const event = e as WebComponentEvent<EventTarget>;
  const path = event.path || (event.composedPath && event.composedPath());
  return path[0];
};

interface EventDetails {
  [key: string]: unknown;
}

const dispatchCustomEvent = (
  selector: string,
  eventType: HeaderFooterEvents | CmsBlocksEvents | CommonEvents,
  detail: EventDetails
): void => {
  const element = document.querySelector(selector);
  if (element && element.dispatchEvent) {
    element.dispatchEvent(
      new CustomEvent(eventType, {
        detail,
      })
    );
  }
};

export const HNeventSetSimplifiedHeader = (isSimplified: boolean = true): void =>
  dispatchCustomEvent('hn-webcomp-header', HeaderFooterEvents.setsimplifiedheader, { simplifiedHeader: isSimplified });

export const HNeventSetFirstTimeLogin = (isFirstTime: boolean = true): void =>
  dispatchCustomEvent('hn-webcomp-header', HeaderFooterEvents.setfirsttimelogin, { firstTimeLogin: isFirstTime });
export const HNeventSetAnonymousHeader = (isAnonymous: boolean = true): void =>
  dispatchCustomEvent('hn-webcomp-header', HeaderFooterEvents.setanonymousheader, { anonymousHeader: isAnonymous });

export const HNeventSetHiddenHeader = (isHidden: boolean = true): void =>
  dispatchCustomEvent('hn-webcomp-header', HeaderFooterEvents.sethiddenheader, { hiddenHeader: isHidden });

export const HNeventSetSimplifiedFooter = (text: string, isSimplified: boolean = true): void =>
  dispatchCustomEvent('hn-webcomp-footer', HeaderFooterEvents.setsimplifiedfooter, {
    simplifiedFooter: isSimplified,
    simplifiedFooterText: text,
  });

export const HNeventSetHiddenFooter = (isHidden: boolean = true): void =>
  dispatchCustomEvent('hn-webcomp-footer', HeaderFooterEvents.sethiddenfooter, {
    hiddenFooter: isHidden,
  });

export const HNeventRefreshVarslinger = (): void => dispatchCustomEvent('hn-webcomp-header', HeaderFooterEvents.refreshvarslinger, {});

export const HNeventSetDriftsmeldingPath = (path: string): void =>
  dispatchCustomEvent('hn-webcomp-driftspanel', HeaderFooterEvents.setdriftsmeldingpath, { path });

export const HNeventRefreshVarslingerOgHendelsesmeny = (): void =>
  dispatchCustomEvent('hn-webcomp-header', HeaderFooterEvents.refreshvarslingeroghendelsesmeny, {});

export const HNeventSetOnShowSignoutbox = (fn: (data: ShowSignOutBoxData) => SessionTimeoutAction): void =>
  dispatchCustomEvent('hn-webcomp-header', HeaderFooterEvents.setonshowsignoutbox, { onShowSignOutBox: fn });

export const HNeventSetUserLoading = (userLoading: boolean): void =>
  dispatchCustomEvent('hn-webcomp-header', HeaderFooterEvents.setuserloading, { userLoading });

export const HNeventSetFromLocation = (fromLocation: History.Location): void =>
  dispatchCustomEvent('hn-webcomp-personvelger', HeaderFooterEvents.setfromlocation, { fromLocation });

export const HNeventSetRedirectToUrlAfterLogin = (redirectToUrlAfterLogin: boolean): void =>
  dispatchCustomEvent('hn-webcomp-personvelger', HeaderFooterEvents.setredirecttourlafterlogin, { redirectToUrlAfterLogin });

export const HNeventSetHistory = (history: History.History<History.LocationState>): void =>
  dispatchCustomEvent('hn-webcomp-personvelger', HeaderFooterEvents.sethistory, { history });

export const HNeventSetLocation = (location: History.Location<History.LocationState>): void =>
  dispatchCustomEvent('hn-webcomp-personvelger', HeaderFooterEvents.setlocation, { location });

export interface VisPersonvelgerDetail extends EventDetails {
  visPersonvelger: boolean;
}

export const HNeventSetVisPersonvelger = (visPersonvelger: boolean): void => {
  setVisPersonvelger(visPersonvelger);
  window.dispatchEvent(
    new CustomEvent<VisPersonvelgerDetail>(HeaderFooterEvents.setvispersonvelger, {
      detail: { visPersonvelger },
    })
  );
};

export const HNeventSetHiddenPromopanel = (isHidden: boolean): void =>
  dispatchCustomEvent('hn-webcomp-cms-block-promopanel', CmsBlocksEvents.setHiddenPromopanel, { hiddenPromopanel: isHidden });

export const HNeventIsMicroFrontendMounted = (webcomponentName: string): void => {
  const webcomponent = document.querySelector(webcomponentName);

  if (webcomponent && webcomponent.dispatchEvent) {
    webcomponent.dispatchEvent(new CustomEvent(CommonEvents.isMounted));
  }
};

export const HNaddEventListener = (
  event: HeaderFooterEvents | CmsBlocksEvents | CommonEvents,
  element: Element | null,
  handler: <T>(EvenCustomEvent: CustomEvent<T>) => void,
  passive = false
): void => {
  useEffect(() => {
    if (element) element.addEventListener(event, handler, passive);

    return function cleanup(): void {
      if (element) element.removeEventListener(event, handler);
    };
  });
};
