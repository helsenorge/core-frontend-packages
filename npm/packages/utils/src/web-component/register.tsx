import React from 'react';
import ReactDOM, { render } from 'react-dom';
import retargetEvents from 'react-shadow-dom-retarget-events';
import { styleInjector, RegisterWebCompSetup } from './helpers';

/** Registrerer/oppretter ny web-komponent.
 *
 * Eksempel:
 * registerWebComp(HeaderWrapper, 'hn-webcomp-header', { events: true, styledComponents: true }, 'hn-webcomp-header-footer-template');
 */

export type Props = { ChildComponent: React.ComponentType<WebcompProps>; name: string; config: Config; templateName: string };

export interface WebcompProps {
  [key: string]: string;
}

export interface Config {
  events: boolean;
  styledComponents: boolean;
}

export default function registerWebComp(
  ChildComponent: React.ComponentType<WebcompProps>,
  name: string,
  config: Config,
  templateName: string
): void {
  class WebComponent extends HTMLElement {
    mountPoint: HTMLElement;
    props: WebcompProps;

    constructor() {
      super();
      this.mountReactApp();
    }

    disconnectedCallback(): void {
      ReactDOM.unmountComponentAtNode(this.mountPoint);
    }

    get value(): WebcompProps {
      return Array.from(this.attributes).reduce((p, c) => {
        p[c.name] = c.value;
        return p;
      }, {}) as WebcompProps;
    }

    eventDispatcher = (event: Event): void => {
      this.dispatchEvent(event);
    };

    // Listen to Events from outside
    subscribeDispatcher = (eventname: string, callback: Function): void => {
      this.addEventListener(eventname, e => {
        return callback(e);
      });
    };

    mountReactApp(): void | null {
      this.props = this.value;
      this.mountPoint = document.createElement('div');

      if (!templateName) {
        console.error('finner ikke template name');
        return null;
      }
      // TEMPLATE
      // Her settes id til templatet
      const tmpl: HTMLTemplateElement = document.getElementById(templateName) as HTMLTemplateElement;
      if (!tmpl) {
        console.error(`Kunne ikke finne html <template> med id: ${templateName} sjekk i index-html at <template> er oppgitt med riktig id`);
        return null;
      }
      // SHADOW-DOM
      // Create an attach to the shadowDOM's root. This is necessary before querying the DOM
      const shadowRoot = this.attachShadow({ mode: 'open' });
      // TEMPLATE + SHADOW-DOM
      // Select the template content and append it to the shadowDOM
      shadowRoot.appendChild(tmpl.content.cloneNode(true));
      shadowRoot.appendChild(this.mountPoint);

      render(
        <RegisterWebCompSetup
          config={config}
          eventDispatcher={this.eventDispatcher}
          subscribeDispatcher={this.subscribeDispatcher}
          mountPoint={this.mountPoint}
        >
          <ChildComponent {...this.props} />
        </RegisterWebCompSetup>,
        this.mountPoint
      );
      styleInjector(this.props, shadowRoot);
      retargetEvents(shadowRoot);
    }
  }

  //her settes navnet til web component elementet
  if (!window.customElements.get(name)) {
    window.customElements.define(name, WebComponent);
  }
}
