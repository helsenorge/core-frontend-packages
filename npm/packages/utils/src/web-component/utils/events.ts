export const refreshUlesteMeldingerEvent = 'hn-refreshUlesteMeldinger';

export const dispatchRefreshUlesteMeldingerEvent = (): void => {
  const webcompheader = document.querySelector('hn-webcomp-header');

  if (webcompheader && webcompheader.dispatchEvent) {
    webcompheader.dispatchEvent(new CustomEvent(refreshUlesteMeldingerEvent));
  }
};
