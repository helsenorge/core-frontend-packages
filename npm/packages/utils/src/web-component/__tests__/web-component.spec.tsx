import React, { createContext, useContext } from 'react';
import { mount } from 'enzyme';

import * as HNPageFunctions from '../../hn-page';
import * as registerWebCompFunctions from '../register';
import * as webcompEvents from '../events';
import WithStore from '../with-store';
import WebCompConsumer from '../consumer';

interface State {
  tester: string;
}
interface Action {
  type: string;
}

describe('Gitt at web component skal instansieres', () => {
  describe('Når withstore brukes', () => {
    it('Så skal elementer som er inni kunne bruke context', () => {
      const reducer = (state: State, action: Action): State => {
        switch (action.type) {
          default: {
            return state;
          }
        }
      };

      const StateContext = createContext<State>({ tester: 'test' });
      const stateDispatch = createContext<React.Dispatch<Action>>(() => {});

      const TestStateDiv: React.FC = () => {
        const state = useContext(StateContext);
        return <div id="testdiv">{state.tester}</div>;
      };

      const state: State = { tester: 'test' };

      const props = {
        StateContext: StateContext,
        DispatchContext: stateDispatch,
        reducer: reducer,
        initialState: state,
      };

      const wrapper = mount(
        <WithStore<State, Action> {...props}>
          <TestStateDiv />
        </WithStore>
      );

      expect(wrapper.find('#testdiv').text()).toBe('test');
    });
  });
});

describe('Gitt at web component skal registreres', () => {
  describe('Når det defineres et navn og en komponent på window objektet', () => {
    it('Så er komponenten registrert på det navnet', async () => {
      const Testwebcomp: React.FC = () => {
        return <div id="testdiv">{'test'}</div>;
      };
      registerWebCompFunctions.registerWebComp(Testwebcomp, 'hn-webcomp-test', { events: true }, 'hn-webcomp-test-template');

      const webCompName = window.customElements.get('hn-webcomp-test');
      await expect(webCompName).toBeTruthy();
    });
  });
});

describe('Gitt at web component skal consumeres', () => {
  describe('Når WebCompConsumer instansieres', () => {
    it('Så fetcher den ressurser fra riktig url og instansierer komponent', async () => {
      jest.spyOn(HNPageFunctions, 'getAssetsUrl').mockImplementation(() => 'assetsUrl');

      const mockSuccessResponse = {
        assets: {
          'entryname.css': '/header-footer/static/css/header-footer.26b379af.css',
          'entryname.js': '/header-footer/static/js/header-footer.3f68cbb6.js',
          'entryname.js.map': '/header-footer/static/js/header-footer.3f68cbb6.js.map',
          'index.html': '/header-footer/index.html',
          'static/css/entryname.26b379af.css.map': '/header-footer/static/css/header-footer.26b379af.css.map',
        },
        entrypoints: {
          entryname: ['static/css/entryname.26b379af.css', 'static/js/entryname.3f68cbb6.js', 'static/js/entryname.3f68cbb6.js.map'],
        },
      };
      const mockJsonPromise = Promise.resolve(mockSuccessResponse);
      const mockFetchPromise = Promise.resolve({
        status: 200,
        response: {},
        json: () => mockJsonPromise,
      });

      jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);
      const isCustomElementRegisteredMock = jest.spyOn(registerWebCompFunctions, 'isCustomElementRegistered');

      const webCompFromMicroFrontend = await mount(
        <WebCompConsumer
          domain="mydomain"
          entryName="entryname"
          componentName={'hn-component'}
          componentProps={{ id: 'testdiv' }}
          includeResetCss
        />
      );

      await expect(global.fetch).toHaveBeenCalledTimes(1);
      await expect(global.fetch).toHaveBeenCalledWith('mydomain/assets.json', {'cache': 'no-cache'});
      await expect(isCustomElementRegisteredMock).toHaveBeenCalledWith('hn-component');
      await expect(webCompFromMicroFrontend.render()).toMatchSnapshot();
    });
  });
});

describe('Gitt at web component events skal kalles', () => {
  describe('Når getEventTarget kalles på en event med path', () => {
    it('Så returnerer den event.path', () => {
      const eventWithPath = ({
        path: ['/eventpath'],
      } as unknown) as Event;
      const result = webcompEvents.getEventTarget(eventWithPath);

      expect(result).toBe('/eventpath');
    });
  });

  describe('Når getEventTarget kalles på en event uten path', () => {
    it('Så kaller den composedPath', () => {
      const composedPathMock = jest.fn().mockImplementation(() => ['/eventcomposedpath']);
      const eventWithoutPath = ({
        composedPath: composedPathMock,
      } as unknown) as Event;
      const result = webcompEvents.getEventTarget(eventWithoutPath);

      expect(composedPathMock).toHaveBeenCalled();
      expect(result).toBe('/eventcomposedpath');
    });
  });
});
