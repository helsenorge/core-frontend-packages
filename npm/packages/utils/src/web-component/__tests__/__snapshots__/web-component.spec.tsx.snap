// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at web component skal consumeres Når WebCompConsumer instansieres Så fetcher den ressurser fra riktig url og instansierer komponent 1`] = `
<hn-component
  id="testdiv"
/>
`;
